ssl/ssl_lib.c

```diff
@@ -4120,13 +4209,28 @@ int ssl_validate_ct(SSL *s)
         goto end;
     }
 
+    issuer = sk_X509_value(s->verified_chain, 1);
     CT_POLICY_EVAL_CTX_set0_cert(ctx, cert);
     CT_POLICY_EVAL_CTX_set0_issuer(ctx, issuer);
     CT_POLICY_EVAL_CTX_set0_log_store(ctx, s->ctx->ctlog_store);
 
     scts = SSL_get0_peer_scts(s);
 
-    if (SCT_LIST_validate(scts, ctx) != 1) {
+    /*
+     * This function returns success (> 0) only when all the SCTs are valid, 0
+     * when some are invalid, and < 0 on various internal errors (out of
+     * memory, etc.).  Having some, or even all, invalid SCTs is not sufficient
+     * reason to abort the handshake, that decision is up to the callback.
+     * Therefore, we error out only in the unexpected case that the return
+     * value is negative.
+     *
+     * XXX: One might well argue that the return value of this function is an
+     * unforunate design choice.  Its job is only to determine the validation
+     * status of each of the provided SCTs.  So long as it correctly separates
+     * the wheat from the chaff it should return success.  Failure in this case
+     * ought to correspond to an inability to carry out its duties.
```

Comment:
CT was (somewhat unfortunately in terms of review cycles) a large and very late addition to the 1.1.0 code base, so the interface did not get much review and is rather new.  So I don't think it is too late to change it.  Indeed now is the best time to change it, later it may have users.

So what do you think?

* Return void, and let the caller examine the status of the list elements?
* Return failure only when out of memory and the like?
* Return < 0 on internal errors (ENOMEM, ...) and 0 when none of the SCTs are valid?

I am rather tempted to say "void" is right, no single outcome represents the status of a list, but a case can be made for either of the other two.  The current return value is sub-optimal.